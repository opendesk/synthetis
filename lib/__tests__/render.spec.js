const readFileSync = require('fs').readFileSync;
const render = require('../render');
const routeFactory = require('../Route').route;
const Fragment = require('../Fragment');
const FragmentBody = require('../FragmentBody');
const Url = require('../Url');

const localBasePath = __dirname + '/fixtures';
const logger = console;
logger.debug = logger.info;

// fetchers

const localDataFetcher = (fragment) => Promise.resolve(new FragmentBody(fragment.localData));

// Note: the fetcher currently has responsibility to handle errors in a non-exception
// manner and obey `required` prop of fragments. This may change in future.
const localFileFetcher = (fragment) => {
  try {
    return Promise.resolve(
      new FragmentBody(readFileSync(fragment.url.resolve({}, false)).toString())
    );
  } catch (e) {
    if (fragment.isRequired) {
      throw e;
    }
    return Promise.resolve(fragment.contentMissingMessage(e));
  }
};

const localFileOrDataFetcher = (fragment) => Promise.resolve(
  fragment.isFile ? localFileFetcher(fragment) : localDataFetcher(fragment)
);

describe('render', () => {
  describe('successful fetch', () => {
    let b, r;

    beforeEach(() => {
      b = Fragment.base({
        data: 'test'
      });
      r = routeFactory({
        baseTemplate: b
      })(console);
    });

    test('should render a basic fragment', async () => {
      const body = await render(r, localDataFetcher);
      expect(body).toMatchSnapshot();
    });
  });

  describe('failed fetch', () => {
    let b, r;

    const e = new Error('Failed to fetch');
    const localDataFetcher = () => Promise.reject(e);

    beforeEach(() => {
      b = Fragment.base({
        data: 'test'
      });
      r = routeFactory({
        baseTemplate: b
      })(console);
    });

    test('should throw', async () => {
      await expect(render(r, localDataFetcher)).rejects.toEqual(e);
    });
  });

  describe('render scenarios', () => {
    let r;

    describe('greeter', () => {
      beforeEach(() => {
        const b = Fragment.base({
          url: Url.localFile(localBasePath, 'greeter.html')
        });
        r = routeFactory({
          baseTemplate: b
        })(console);
      });

      test('should render a basic fragment', async () => {
        const body = await render(r, localFileFetcher);
        expect(body).toMatchSnapshot();
      });
    });

    describe('base template with simple fragment inject', () => {
      beforeEach(() => {
        const f = Fragment.html({
          url: Url.localFile(localBasePath, 'greeter.html')
        });
        const b = Fragment.base({
          url: Url.localFile(localBasePath, 'simpleTemplate.html')
        });
        r = routeFactory({
          baseTemplate: b,
          fragments: {
            greeter: f
          }
        })(logger);
      });

      test('should render fragment', async () => {
        const body = await render(r, localFileFetcher);
        expect(body).toMatchSnapshot();
      });
    });

    describe('base template with embedded fragment with injected data', () => {
      beforeEach(() => {
        const f = Fragment.json({
          data: {
            myData: 1234
          }
        });
        const b = Fragment.base({
          url: Url.localFile(localBasePath, 'simpleTemplateWithEmbed.html')
        });
        r = routeFactory({
          baseTemplate: b,
          fragments: {
            dataModel: f
          }
        })(logger);
      });

      test('should render fragment', async () => {
        const body = await render(r, localFileOrDataFetcher);
        expect(body).toMatchSnapshot();
      });
    });

    describe('base template with repeat embedded fragment with injected data', () => {
      beforeEach(() => {
        const dataModel = Fragment.json({
          data: JSON.parse(readFileSync(localBasePath + '/basicData.json').toString())
        });
        const b = Fragment.base({
          url: Url.localFile(localBasePath, 'templateWithRepeatEmbed.html')
        });
        r = routeFactory({
          baseTemplate: b,
          fragments: {
            dataModel
          }
        })(logger);
      });

      test('should render fragment', async () => {
        const body = await render(r, localFileOrDataFetcher);
        expect(body).toMatchSnapshot();
      });
    });

    describe('base template with repeat embedded fragment with no data', () => {
      beforeEach(() => {
        const f = Fragment.json({
          data: undefined
        });
        const b = Fragment.base({
          url: Url.localFile(localBasePath, 'templateWithRepeatEmbed.html')
        });
        r = routeFactory({
          baseTemplate: b,
          fragments: {
            dataModel: f
          }
        })(logger);
      });

      test('should render fragment', async () => {
        const body = await render(r, localFileOrDataFetcher);
        expect(body).toMatchSnapshot();
      });
    });

    describe('base template with nested fragments with injected data', () => {
      beforeEach(() => {
        const anotherTemplate = Fragment.html({
          url: Url.localFile(localBasePath, 'anotherTemplate.html')
        });
        const nestedFragment = Fragment.html({
          url: Url.localFile(localBasePath, 'simpleTemplateWithEmbed.html')
        });
        const dataModel = Fragment.json({
          data: {
            myData: 1234
          }
        });
        const b = Fragment.base({
          url: Url.localFile(localBasePath, 'simpleTemplate.html')
        });
        r = routeFactory({
          baseTemplate: b,
          fragments: {
            greeter: anotherTemplate,
            nestedFragment,
            dataModel
          }
        })(logger);
      });

      test('should render fragment', async () => {
        const body = await render(r, localFileOrDataFetcher);
        expect(body).toMatchSnapshot();
      });
    });
  });

  describe('rendering error scenarios', () => {
    describe('base template with embedded fragment with missing data', () => {
      beforeEach(() => {
        const f = Fragment.json({
          data: undefined
        });
        const b = Fragment.base({
          url: Url.localFile(localBasePath, 'simpleTemplateWithEmbed.html')
        });
        r = routeFactory({
          baseTemplate: b,
          fragments: {
            dataModel: f
          }
        })(logger);
      });

      test('should render fragment', async () => {
        const body = await render(r, localFileOrDataFetcher);
        expect(body).toMatchSnapshot();
      });
    });

    describe('base template with repeat embedded fragment with missing data', () => {
      const f = Fragment.json({
        data: {
          data: {
            array: [undefined, undefined]
          }
        }
      });
      beforeEach(() => {
        const b = Fragment.base({
          url: Url.localFile(localBasePath, 'templateWithRepeatEmbed.html')
        });
        r = routeFactory({
          baseTemplate: b,
          fragments: {
            dataModel: f
          }
        })(logger);
      });

      test('should render fragment', async () => {
        const body = await render(r, localFileOrDataFetcher);
        expect(body).toMatchSnapshot();
      });
    });

    describe('missing local base path', () => {
      beforeEach(() => {
        const b = Fragment.base({});
        r = routeFactory({
          baseTemplate: b
        })(console);
      });

      test('should throw', async () => {
        await expect(render(r, localDataFetcher)).rejects.toMatchSnapshot();
      });
    });

    describe('base template with not required missing fragment', () => {
      beforeEach(() => {
        const f = Fragment.html({
          url: Url.localFile(localBasePath, 'foobar.html')
        });
        const b = Fragment.base({
          url: Url.localFile(localBasePath, 'simpleTemplate.html')
        });
        r = routeFactory({
          baseTemplate: b,
          fragments: {
            greeter: f
          }
        })(logger);
      });

      test('should render fragment and show error message', async () => {
        const body = await render(r, localFileFetcher);
        expect(body).toMatchSnapshot();
      });
    });

    describe('base template with unknown fragment name', () => {
      beforeEach(() => {
        const b = Fragment.base({
          url: Url.localFile(localBasePath, 'simpleTemplate.html')
        });
        r = routeFactory({
          baseTemplate: b
        })(logger);
      });

      test('should throw and not render', async () => {
        await expect(render(r, localFileFetcher)).rejects.toMatchSnapshot();
      });
    });


    describe('base template with required missing fragment', () => {
      beforeEach(() => {
        const f = Fragment.html({
          url: Url.localFile(localBasePath, 'foobar.html'),
          required: true
        });
        const b = Fragment.base({
          url: Url.localFile(localBasePath, 'simpleTemplate.html')
        });
        r = routeFactory({
          baseTemplate: b,
          fragments: {
            greeter: f
          }
        })(logger);
      });

      test('should throw and not render', async () => {
        await expect(render(r, localFileFetcher)).rejects.toBeDefined();
      });
    });
  });
});
