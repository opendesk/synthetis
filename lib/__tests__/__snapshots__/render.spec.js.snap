// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render render scenarios base template with embedded fragment with injected data should render fragment 1`] = `
FragmentBody {
  "_body": "<p>template</p><p>embed 1234</p>",
  "_contentType": "text/html",
}
`;

exports[`render render scenarios base template with nested fragments with injected data should render fragment 1`] = `
FragmentBody {
  "_body": "<p>template</p><p>another template<span><p>template</p><p>embed 1234</p></span></p>",
  "_contentType": "text/html",
}
`;

exports[`render render scenarios base template with repeat embedded fragment with injected data should render fragment 1`] = `
FragmentBody {
  "_body": "<p>template</p><p>embed one</p><p>embed two</p><p>embed three</p>",
  "_contentType": "text/html",
}
`;

exports[`render render scenarios base template with repeat embedded fragment with no data should render fragment 1`] = `
FragmentBody {
  "_body": "<p>template</p>",
  "_contentType": "text/html",
}
`;

exports[`render render scenarios base template with simple fragment inject should render fragment 1`] = `
FragmentBody {
  "_body": "<p>template</p><p>hi</p>",
  "_contentType": "text/html",
}
`;

exports[`render render scenarios greeter should render a basic fragment 1`] = `
FragmentBody {
  "_body": "<p>hi</p>",
  "_contentType": "text/html",
}
`;

exports[`render rendering error scenarios base template with embedded fragment with missing data should render fragment 1`] = `
FragmentBody {
  "_body": "<p>template</p><p>Sorry this content could not be rendered. Please try again.</p>",
  "_contentType": "text/html",
}
`;

exports[`render rendering error scenarios base template with not required missing fragment should render fragment and show error message 1`] = `
FragmentBody {
  "_body": "<p>template</p><p>Sorry this content could not be fetched. Please try again.</p>",
  "_contentType": "text/html",
}
`;

exports[`render rendering error scenarios base template with repeat embedded fragment with missing data should render fragment 1`] = `
FragmentBody {
  "_body": "<p>template</p><p>Sorry this content could not be rendered. Please try again.</p>",
  "_contentType": "text/html",
}
`;

exports[`render rendering error scenarios base template with unknown fragment name should throw and not render 1`] = `[Error: A source fragment does not exist with name greeter]`;

exports[`render rendering error scenarios missing local base path should throw 1`] = `[Error: A fragment body is undefined]`;

exports[`render successful fetch should render a basic fragment 1`] = `
FragmentBody {
  "_body": "test",
  "_contentType": "text/html",
}
`;
